rpc/query_server.go

```diff
@@ -749,36 +756,44 @@ func (s *QueryServer) EthGetTransactionReceipt(hash eth.Data) (resp eth.JsonTxRe
 	if err != nil {
 		return resp, err
 	}
+
 	txReceipt, err := r.GetReceipt(snapshot, txHash)
 	if err != nil && errors.Cause(err) != common.ErrTxReceiptNotFound {
-		return resp, err
+		// return empty response if cannot find hash
+		resp.Status = eth.EncInt(int64(StatusTxFail))
+		return resp, nil
 	}
 	if err != nil {
-		txObj, err := query.GetTxByTendermintHash(s.BlockStore, txHash)
+		resp, err = getReceiptByTendermintHash(snapshot, s.BlockStore, r, txHash)
 		if err != nil {
-			return resp, errors.Wrapf(err, "cannot find tx hash %v", txHash)
+			// return empty response if cannot find hash
+			resp.Status = eth.EncInt(int64(StatusTxFail))
+			return resp, nil
 		}
-		return eth.TxObjToReceipt(txObj), nil
+		return resp, err
 	}
-
-	// accessing the TM block store might take a while and we don't need the snapshot anymore
 	snapshot.Release()
 
-	if len(txReceipt.Logs) > 0 {
-		height := int64(txReceipt.BlockNumber)
-		var blockResult *ctypes.ResultBlock
-		blockResult, err := s.BlockStore.GetBlockByHeight(&height)
-		if err != nil {
-			return resp, err
-		}
-		timestamp := blockResult.Block.Header.Time.Unix()
-
-		for i := 0; i < len(txReceipt.Logs); i++ {
-			txReceipt.Logs[i].BlockTime = timestamp
-		}
+	height := int64(txReceipt.BlockNumber)
+	blockResult, err := s.BlockStore.GetBlockByHeight(&height)
+	if err != nil {
+		// return empty response if cannot find hash
+		resp.Status = eth.EncInt(int64(StatusTxFail))
+		return resp, nil
+	}
+	if int32(len(blockResult.Block.Data.Txs)) <= txReceipt.TransactionIndex {
+		// return empty response if cannot find hash
+		resp.Status = eth.EncInt(int64(StatusTxFail))
+		return resp, nil
```

Comment:
Here too it probably makes more sense to return an error rather than a fake receipt.